x/cdp/abci.go

```diff
@@ -42,6 +47,22 @@ func BeginBlocker(ctx sdk.Context, req abci.RequestBeginBlock, k Keeper) {
 			),
 		)
 	}
+	distTimeElapsed := sdk.NewInt(ctx.BlockTime().Unix() - previousDistTime.Unix())
+	if distTimeElapsed.GTE(params.SavingsDistributionFrequency) {
+		for _, dp := range params.DebtParams {
+			err := k.ApplySavingsRate(ctx, dp.Denom)
+			if err != nil {
+				ctx.EventManager().EmitEvent(
+					sdk.NewEvent(
+						EventTypeBeginBlockerFatal,
+						sdk.NewAttribute(sdk.AttributeKeyModule, fmt.Sprintf("%s", ModuleName)),
+						sdk.NewAttribute(types.AttributeKeyError, fmt.Sprintf("%s", err)),
+					),
+				)
+			}
+		}
+		k.SetPreviousSavingsDistribution(ctx, ctx.BlockTime())
+	}
```

Comment:
Should paying out savings run before running surplus and debt auctions?

Reply:
It should not. 
As currently implemented, surplus checks if (surplus * (1-savings_rate)) > surplus_threshold. Imagine if the surplus_threshold was 5, the current surplus is 100, and the savings_rate is 0.95. If savings rate distribution went first, 95 of the surplus would be paid out, then the surplus auction would check (5 * 0.05) = 0.25. 

Point taken on your suggestion for moving rewards to a separate module account. I agree it's hard to reason about as implemented.